test/functional/README.md

```diff
@@ -0,0 +1,285 @@
+```bash
+$ source testlib.bash
+```
+
+2. If necessary, create a directory to group the new test script with other test scripts of the same theme.  
+```bash
+$ mkdir <some_test_theme>
+$ cd <some_test_theme>
+```
+
+3. Create the new test script.  
+```bash
+$ generate_test <test_name>
+```
+
+A new test script will be generated in the current directory.  
+```bash
+#!/usr/bin/env bats
+
+load "../testlib"
+
+setup() {
+
+	create_test_environment "$TEST_NAME"
+	# create_bundle -n <bundle_name> -f <file_1>,<file_2>,<file_N> "$TEST_NAME"
+
+}
+
+teardown() {
+
+	destroy_test_environment "$TEST_NAME"
+
+}
+
+@test "<test description>" {
+
+	run sudo sh -c "$SWUPD <swupd_command> $SWUPD_OPTS <command_options>"
+	# <validations>
+
+}
+```
+
+4. The setup() function contains commands that should be executed before every test in the test script. This is usually the best place for creating test objects that are pre-requists for all tests in the script. The test library provides many [fixtures](#test-fixtures) to create all these pre-requisits easily.
+
+5. The teardown() function contains commands that should be executed after every test in the test script. It is the most common place for cleaning up test resources created for and by tests.
+
+6. Every test within the test file starts with the @test identifier followed by a description of the test. Add one or more tests to the script as needed. 
+
+7. From within each test, execute the command you want to validate by using the `run` helper.
+
+8. Validate that the program under test performed as expected by using one or more of the [assertions](#assertions) provided by the test library.
+<br/>
+
+### Test Principles
+
+* Every test needs to run in its own test environment so they don't interfere with each other.
+* Every test should clean up after itself.
+* Tests should be atomic and test only one thing per test.
+* Tests should be independent from each other, one should not need another one in order to work.
+* Test files from the same "theme" should be grouped within the theme directory. Tests from the same type can optionally be added to the same test file, for example:
+  * bundle-add  # A theme
+    * add-single-bundle.bats
+      * @test "add a bundle"
+      * @test "add a bundle that has only a directory"
+      * @test "add a bundle that has many files"
+      * @test "add a bundle that is already installed"
+      * @test "add a bundle that is invalid"
+      * etc.
+    * add-multiple-bundles.bats
+      * @test "add multiple bundles"
+      * @test "add multiple bundles, one valid, one invalid"
+      * etc.
+  * bundle-remove  # Another theme
+  * etc.
+<br/>
+
+### Running Tests
+To run a specific test script locally:  
+```bash
+$ bats <theme_directory>/<test_script>.bats
+```  
+
+To run all tests from a theme locally:  
+```bash
+$ bats <theme_directory>
+```  
+
+To run all tests locally:
```

Comment:
I had not included that in the documentation yet because that is actually not the case yet, tests written with this library have not been added to the Makefile yet. I wanted to discuss with you if you really wanted to keep listing each test individually in the Makefile.am, I don't see the purpose in adding each one individually when we can just run something that would "discover" all the tests automatically, so people wouldn't have to worry about that with every new test. I like the idea of running `make check` as a way of running them, but I guess not the way the Makefile is including all those tests. Is there a specific reason why they are being added one by one?

Reply:
> we can just run something that would "discover" all the tests automatically

This would be great. Until this is done please add them to the Makefile.am so they are actually run (or just add the discovery).